### PR TITLE
Changefeed

### PR DESCRIPTION
**Pull request by xiafu-msft**

### CODE REVIEW

**Comment by seanmcc-msft** [comment]
Is this the entire Change Feed implementation?

**Comment by xiafu-msft** [comment]
> Is this the entire Change Feed implementation?

It's not! Only a draft pr!

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/__init__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# license information.
+# --------------------------------------------------------------------------
+
+from azure.core.paging import ItemPaged
+from ._change_feed_client import ChangeFeedClient
+
+
+__all__ = [
+    'ChangeFeedClient',
+    'ItemPaged'
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/__init__.py:13]: We don't need to expose `ItemPaged` here - users are unlikely to need this type, and if they do they can import it from azure-core directly.

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_change_feed_client.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    :param credential:
+        The credentials with which to authenticate. This is optional if the
+        account URL already has a SAS token. The value can be a SAS token string, an account
+        shared access key, or an instance of a TokenCredentials class from azure.identity.
+        If the URL already has a SAS token, specifying an explicit credential will take priority.
+    :keyword str api_version:
+        The Storage API version to use for requests. Default value is '2019-12-12'.
+        Setting to an older version may result in reduced feature compatibility.
+
+        .. versionadded:: 12.2.0
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_change_feed_client.py:34]: We don't need the `versionadded` tag here :)

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_change_feed_client.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                :language: python
+                :dedent: 8
+                :caption: Creating the BlobServiceClient using account_key as credential.
+        """
+        account_url, secondary, credential = parse_connection_str(conn_str, credential, 'blob')
+        if 'secondary_hostname' not in kwargs:
+            kwargs['secondary_hostname'] = secondary
+        return cls(account_url, credential=credential, **kwargs)
+
+    def get_changes(self, start_time=None, end_time=None, **kwargs):
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_change_feed_client.py:89]: I'm wondering if this should be `list_changes`?

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+# --------------------------------------------------------------------------
+# pylint: disable=too-few-public-methods, too-many-instance-attributes
+# pylint: disable=super-init-not-called, too-many-lines
+import collections
+import json
+from datetime import datetime
+from io import BytesIO
+
+from avro.datafile import DataFileReader
+from avro.io import DatumReader
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py:14]: I thought we were using the new internal avro parser?

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+            if self.end_time and self._is_later_than_end_time(segment_path):
+                self.unprocessed_segment_paths = []
+                return None
+            return Segment(self.client, segment_path, page_size, segment_cursor)
+        return None
+
+    @staticmethod
+    def _parse_datetime_from_segment_path(segment_path):
+        path_tokens = segment_path.name.split("/")
+        return datetime(int(path_tokens[2]), int(path_tokens[3]), int(path_tokens[4]), int(path_tokens[5][:2]))
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py:149]: This could be moved out of the class into a utility function?

File: sdk/storage/azure-storage-blob-changefeed/setup.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        'Programming Language :: Python :: 3.6',
+        'Programming Language :: Python :: 3.7',
+        'Programming Language :: Python :: 3.8',
+        'License :: OSI Approved :: MIT License',
+    ],
+    zip_safe=False,
+    packages=find_packages(exclude=[
+        # Exclude packages that will be covered by PEP420 or nspkg
+        'azure',
+        'azure.storage',
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/setup.py:90]: Do we intend to ship an async version? Or will that be in a later preview?

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py (hunk 4 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py]: Yeah, at that time internal parser wasn't ready, it's switched to internal parser now!

File: sdk/storage/azure-storage-blob-changefeed/setup.py (hunk 6 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob-changefeed/setup.py:90]: The async version will be in the next preview probably

File: sdk/storage/azure-storage-blob-changefeed/setup.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+import os
+import re
+
+from setuptools import setup, find_packages
+
+
+# Change the PACKAGE_NAME only to change folder and different name
+PACKAGE_NAME = "azure-storage-blob-changefeed"
+NAMESPACE_NAME = "azure.storage.blobchangefeed"
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/setup.py:18]: I feel like the namespace should also be "azure.storage.blob.changefeed"
Is there a reason that it isn't?
I'll discuss it with Johan as well.

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_deserialize.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+_SERVICE_PARAMS = {
+    "blob": {"primary": "BlobEndpoint", "secondary": "BlobSecondaryEndpoint"},
+}
+
+
+def parse_connection_str(conn_str, credential, service):
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_deserialize.py:12]: We already have a dependency on Storage-Blobs being installed - maybe we could just call into that?

File: sdk/storage/azure-storage-blob-changefeed/setup.py (hunk 7 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob-changefeed/setup.py:18]: We need to declare another namespace package if we want to extend this namespace azure.storage.blob 
This is one of the reasons that we are using azure.storage.filedatalake instead of azure.storage.file.datalake for now.

Comment (annatisch) [reply sdk/storage/azure-storage-blob-changefeed/setup.py:18]: In this case we shouldn't need a namespace package because we have a dependency on blobs which _is_ the namespace.

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+
+    :ivar int results_per_page:
+        The maximum number of results retrieved per API call.
+    :ivar dict continuation_token:
+        The continuation token to retrieve the next page of results.
+    :ivar current_page:
+        The current page of listed results.
+    :vartype current_page: list(dict)
+
+    :param ~azure.storage.blob.ContainerClient:
```
Comment (rakshith91) [inline sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py:36]: this can be blob.aio.ContainerClient too ;)

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+                    self._is_earlier_than_start_time(self.unprocessed_segment_paths[0]):
+                self.unprocessed_segment_paths.popleft()
+
+        # if change_feed_cursor is specified, start from the specified segment
+        if cf_cursor:
+            while self.unprocessed_segment_paths[0].name != cf_cursor['segment_path']:
+                self.unprocessed_segment_paths.popleft()
+
+        self.current_segment = self._get_next_segment(self.page_size,
+                                                      segment_cursor=cf_cursor['segment_cursor'] if cf_cursor else None)
```
Comment (rakshith91) [inline sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py:139]: funny indentation

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+        except StopIteration:
+            pass
+
+        start_point = self._point
+
+        # EOF
+        if self._point + size > self._download_offset:
+            self._point = self._download_offset
+        else:
+            self._point += size
```
Comment (rakshith91) [inline sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py:392]:
```suggestion
        self._point = min(self._point+size, self._download_offset)
```

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_deserialize.py (hunk 8 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_deserialize.py:12]: good catch!

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py (hunk 9 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py:36]: good catch!

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py (hunk 10 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob-changefeed/azure/storage/blobchangefeed/_models.py:139]: This is auto formatted by Ctrl+Alt+L 😀

File: sdk/storage/azure-storage-blob/azure/storage/blob/__init__.py [hunk 12]
```diff
@@ -3,6 +3,7 @@
 # Licensed under the MIT License. See License.txt in the project root for
 # license information.
 # --------------------------------------------------------------------------
+__path__ = __import__('pkgutil').extend_path(__path__, __name__)  # type: str
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/__init__.py:6]: Hi Anna @annatisch 
After changed the namesapce to azure.storage.blob.changefeed, I had to add this line. I was able to run the test locally, while tox.ini is failing, do you have any idea about this?

Comment (lmazuel) [reply sdk/storage/azure-storage-blob/azure/storage/blob/__init__.py:6]: This file is a production file (it ends up on a customer machine), so we must not add this line, the rest is CI problems for us.

Comment (annatisch) [reply sdk/storage/azure-storage-blob/azure/storage/blob/__init__.py:6]: Thanks @lmazuel! @xiafu-msft - lets remove this line and figure out how to resolve CI :)

Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/__init__.py:6]: Thanks @lmazuel, @annatisch !
While if this line is removed, currently I haven't figured out how to import from azure.storage.blob, because the blob/\__init\__.py under azure./storage/blob/changefeed doesn't define anything.

File: sdk/storage/azure-storage-blob-changefeed/setup.py (hunk 7 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob-changefeed/setup.py]: We are discussing this in another open comment, so I will resolve this comment!

File: sdk/storage/azure-storage-blob/azure/storage/blob/__init__.py (hunk 12 above)
Comment (annatisch) [reply sdk/storage/azure-storage-blob/azure/storage/blob/__init__.py:6]: We use this shared namespace pattern in other places that you can use as an example. Take a look at how they handle setup.py, directory structure, dev env set up etc
- Event Hubs extensions. This package takes a hard dependency on the azure-eventhubs sdk and shares it's namespace.
https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/eventhub/azure-eventhub-checkpointstoreblob
- Telemetry extensions. This package takes a hard dependency on azure-core and shares it's namespace:
https://github.com/Azure/azure-sdk-for-python/tree/master/sdk/core/azure-core-tracing-opentelemetry

You can keep this line in your local environment if it helps you run tests locally etc - however please remove it from checked in code.

File: sdk/storage/azure-storage-blob-changefeed/setup.py [hunk 13]
```diff
@@ -88,6 +88,7 @@
         # Exclude packages that will be covered by PEP420 or nspkg
         'azure',
         'azure.storage',
+        'azure.storage.blob',
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob-changefeed/setup.py:91]: it was working when running tox locally...

File: sdk/storage/azure-storage-blob-changefeed/MANIFEST.in [hunk 14]
```diff
@@ -0,0 +1,6 @@
+include *.md
+include azure/__init__.py
+include azure/storage/__init__.py
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/MANIFEST.in:3]: You also need to add:
```
include azure/storage/blob/__init__.py
```

File: sdk/storage/azure-storage-blob-changefeed/mypy.ini [hunk 15]
```diff
@@ -0,0 +1,14 @@
+[mypy]
+python_version = 3.6
+warn_return_any = True
+warn_unused_configs = True
+ignore_missing_imports = True
+
+# Per-module options:
+
+[mypy-azure.storage.blob._generated.*]
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/mypy.ini:9]: Is this necessary?

File: sdk/storage/azure-storage-blob-changefeed/samples/change_feed_samples.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+        # [END list_events_by_page]
+
+    def list_all_events(self):
+        # [START list_all_events]
+        cf_client = ChangeFeedClient("http://{}.blob.core.windows.net".format(self.ACCOUNT_NAME),
+                                     credential=self.ACCOUNT_KEY)
+        change_feed = cf_client.list_changes()
+
+        # print all events
+        events = list(change_feed)
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/samples/change_feed_samples.py:64]: We should remove this line - we don't want to encourage pre-downloading the generator, as for some people this could be really slow.

Comment (annatisch) [reply sdk/storage/azure-storage-blob-changefeed/samples/change_feed_samples.py:64]: In fact we can remove the `list()` casting in each of the samples.

File: sdk/storage/ci.yml [hunk 17]
```diff
@@ -46,7 +46,9 @@ stages:
     ServiceDirectory: storage
     Artifacts:
     - name: azure_storage_blob
-      safeName: azurestorageblob
```
Comment (annatisch) [inline sdk/storage/ci.yml:49]: Why was this removed?

File: sdk/storage/azure-storage-blob-changefeed/README.md [hunk 18]
```diff
@@ -0,0 +1,190 @@
+# Azure Storage Blob ChangeFeed client library for Python
+
+Overview
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/README.md:3]: We can either remove this or make it a subheader.

File: sdk/storage/azure-storage-blob-changefeed/README.md [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+
+This preview package for Python enable users to get change feed events by page, get all change feed events, get events in a time range, list events with a continuation token
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/README.md:5]: Just some grammar tweaks:
"This preview package for Python enables users to get blob change feed events. These events can be lazily generated, iterated by page, retrieved for a specific time interval, or iterated from a specific continuation token."

File: sdk/storage/azure-storage-blob-changefeed/README.md [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+# if using an existing resource group, skip this step
+az group create --name my-resource-group --location westus2
+
+# Create the storage account
+az storage account create -n my-storage-account-name -g my-resource-group --hierarchical-namespace true
+```
+
+### Authenticate the client
+
+Interaction with Blob ChangeFeed client starts with an instance of the ChangeFeed class. You need an existing storage account, its URL, and a credential to instantiate the client object.
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/README.md:42]: "ChangeFeedClient class"

File: sdk/storage/azure-storage-blob-changefeed/README.md [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+
+service = ChangeFeedClient(account_url="https://<my-storage-account-name>.dfs.core.windows.net/", credential=credential)
+```
+
+## Key concepts
+
+#### Clients
+
+The Blob ChangeFeed SDK provides one client:
+ChangeFeedClient: get change feed events by page, get all change feed events, get events in a time range, start listing events with a continuation token
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/README.md:70]: This section isn't rendering quite right - I think it needs a bullet point:
https://github.com/Azure/azure-sdk-for-python/blob/bae3f4aefd644fb974e016ab3fb355b19ad15316/sdk/storage/azure-storage-blob-changefeed/README.md#clients

Also description needs a capital and a final period.

File: sdk/storage/azure-storage-blob-changefeed/README.md [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+from azure.storage.blob.changefeed import ChangeFeedClient
+
+cf_client = ChangeFeedClient("http://{}.blob.core.windows.net".format("YOUR_ACCOUNT_NAME"),
+                             credential="Your_ACCOUNT_KEY")
+start_time = datetime(2020, 1, 6)
+end_time = datetime(2020, 3, 4)
+change_feed = cf_client.list_changes(start_time=start_time, end_time=end_time)
+
+# print first page of events
+events = list(change_feed)
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob-changefeed/README.md:104]: Please remove all the `list()` casting from the samples.

File: sdk/storage/azure-storage-blob-changefeed/README.md (hunk 18 above)
Comment (rakshith91) [reply sdk/storage/azure-storage-blob-changefeed/README.md:3]: @xiafu-msft here are the list of sections you must have in the read me
https://review.docs.microsoft.com/en-us/help/contribute-ref/contribute-ref-how-to-document-sdk?branch=master#readme

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blob/changefeed/_change_feed_client.py [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+        ):
+        # type: (...) -> None
+        self._blob_service_client = BlobServiceClient(account_url, credential, **kwargs)
+
+    @classmethod
+    def from_connection_string(
+            cls, conn_str,  # type: str
+            credential=None,  # type: Optional[Any]
+            **kwargs  # type: Any
+        ):  # type: (...) -> BlobServiceClient
```
Comment (rakshith91) [inline sdk/storage/azure-storage-blob-changefeed/azure/storage/blob/changefeed/_change_feed_client.py:55]:
```suggestion
        ):  # type: (...) -> ChangeFeedClient
```

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blob/changefeed/_change_feed_client.py [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+                :dedent: 8
+                :caption: Creating the BlobServiceClient using account_key as credential.
+        """
+        account_url, secondary, credential = parse_connection_str(conn_str, credential, 'blob')
+        if 'secondary_hostname' not in kwargs:
+            kwargs['secondary_hostname'] = secondary
+        return cls(account_url, credential=credential, **kwargs)
+
+    def list_changes(self, **kwargs):
+        # type: (Optional[datetime], Optional[datetime], **Any) -> ItemPaged[BlobProperties]
```
Comment (rakshith91) [inline sdk/storage/azure-storage-blob-changefeed/azure/storage/blob/changefeed/_change_feed_client.py:84]: ItemPaged[BlobProperties] is this correct?

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blob/changefeed/_models.py [hunk 25]
```diff
@@ -0,0 +1,407 @@
...
+                                                      segment_cursor=cf_cursor['segment_cursor'] if cf_cursor else None)
+
+    def _get_next_segment(self, page_size, segment_cursor=None):
+        if self.unprocessed_segment_paths:
+            segment_path = self.unprocessed_segment_paths.popleft()
+            if self.end_time and self._is_later_than_end_time(segment_path):
+                self.unprocessed_segment_paths = []
+                return None
+            return Segment(self.client, segment_path, page_size, segment_cursor)
+        return None
```
Comment (rakshith91) [inline sdk/storage/azure-storage-blob-changefeed/azure/storage/blob/changefeed/_models.py:147]: why do we need the last return None?

File: sdk/storage/azure-storage-blob-changefeed/mypy.ini (hunk 15 above)
Comment (rakshith91) [reply sdk/storage/azure-storage-blob-changefeed/mypy.ini:9]: i believe yes when using the latest autorest.

File: sdk/storage/azure-storage-blob-changefeed/MANIFEST.in (hunk 14 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob-changefeed/MANIFEST.in:3]: good catch

File: sdk/storage/azure-storage-blob-changefeed/mypy.ini (hunk 15 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob-changefeed/mypy.ini:9]: @annatisch I think you are right, we don't need this since there's no _generated folder in changefeed package.
while why do we need [mypy-azure.core.*] will mypy throw error in dependency package?

File: sdk/storage/azure-storage-blob-changefeed/README.md (hunk 19 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob-changefeed/README.md:5]: this sounds much better!

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blob/changefeed/_change_feed_client.py (hunk 24 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob-changefeed/azure/storage/blob/changefeed/_change_feed_client.py:84]: good catch, it should be dict

File: sdk/storage/azure-storage-blob-changefeed/azure/storage/blob/changefeed/_models.py (hunk 25 above)
Comment (xiafu-msft) [reply sdk/storage/azure-storage-blob-changefeed/azure/storage/blob/changefeed/_models.py:147]: `self.current_segment = self._get_next_segment(.....)`
because if there's no Segment we need to mark self.current_segment = None

**Comment by rakshith91** [comment]
you should be able to resolve the lint error once you rebase
